Update Module1 through Module5 with audio that auto plays when we cli…

ScriabinOp8No12 opened the pull request:
…ck a button to navigate to that page, also have a button 'play audio' on each page (audio is stored on cloudinary cdn and we use a url link to play that audio in our code). Changed various text to match the audio, and some board diagrams.

Also, I've updated the yarn version from 1.22.18 to 1.22.21, will that pose any problems?

ScriabinOp8No12 committed:
Update Module1 through Module5 with audio that auto plays when we click a button to navigate to that page, also have a button 'play audio' on each page as well (audio is stored on cloudinary cdn and we use a url link to play that audio in our code), also changed various text and some board diagrams to match what we wanted

ScriabinOp8No12 commented:
TLDR: Audio auto-plays on modules 1-5, a "play audio" button was added, text and diagrams were changed.

Edit: 7 more commits include: 

1. Audio has been trimmed to have less pauses, and are now mp3 files instead of wav files
2. Text shows up right away instead of animating in
3. Easy bot +6 and option to take 6 stone handicap removed from list of options
4. Added text under buttons on the lessons page describing what each lesson teaches

anoek commented on src/views/Lessons/Module1.tsx:
```diff
@@ -24,14 +24,52 @@ import { openPopup } from "PopupDialog";
+                key="audioElement"
+                ref={this.audioRef}
+                style={{ visibility: "hidden" }}
+                autoPlay={true} // This line auto plays the audio when we click the next button to navigate to the next page
```
FYI autoPlay doesn't work before you've interacted with the page in (most? all?) modern browsers.

This is one of the reasons why I've used ["howler"](https://howlerjs.com/) in the past to help manage that. You can do some fancy sprite sheet stuff to preload audio assets as well, as well as change the volume easily, which is something we might want to consider doing with these samples as well. Here's an example of how I used it in the kidsgo repo: https://github.com/online-go/kidsgoserver.com/blob/devel/src/lib/kidsgo-sfx.ts

anoek replied:
To follow up on this, the howler library will deal with auto playing once the user has "unlocked" auto-play by interacting with the site in some way, as opposed to the `autoPlay` attribute which will work if you've already interacted with the page, but won't work if they are linked to the site externally or reloaded the page or something like that.

ScriabinOp8No12 replied:
Yeah, I was having trouble with autoPlay, thanks for the example.  

Are you saying howler can allow us to autoplay the audio once the user has unlocked auto-play for the first time, and then a page refresh or changing the url manually would still play the audio?  

Also, if we are linked to the site externally and the user hasn't interacted with the page yet, will howler actually play the audio right away, or that still requires an initial user interaction?

anoek commented on src/views/Lessons/Module1.tsx:
```diff
@@ -24,14 +24,52 @@ import { openPopup } from "PopupDialog";
 class Page1 extends Module1 {
+    constructor() {
+        super(
+            "https://res.cloudinary.com/dn8rdavoi/video/upload/v1707854748/kids-go-server-COMBINED-audio-slices/slice_audio1_nkvkon.wav",
```
Might want to consider converting these to mp3/ogg/webm for audio, these wav files are pretty big

ScriabinOp8No12 replied:
Yeah, I was worried we'd lose audio quality if I converted them to mp3 format.  Will the larger size cause significant performance dips even though they are stored on a CDN?

anoek replied:
> Yeah, I was worried we'd lose audio quality if I converted them to mp3 format 

Nah you won't be able to tell the difference.

> Will the larger size cause significant performance dips even though they are stored on a CDN? 

On a wired connection (cable, fiber, dsl)  probably not realistically. If someone is out and about and is using mobile data and they don't have the best reception, the difference could be notable, although if the player is playing while it's streaming, which it probably does, probably not even then. So maybe it's all just fine as is, I'm probably showing my age a bit squabbling over a few MB.

Still though, that first slice 81 we're at 1.34MB, so would take a bit over a second to download on a 10Mbs connection (probably closer to 2s when you add in the latency for establishing the connection and ramping the TCP speed), the mp3 for that is 112KB and ogg at 70KB, so those would be 1/10th of a second and 1/20th of a second respectively (plus whatever connection latency), so would feel pretty instant. That said, browsers probably will start actually playing the audio before the whole file is done downloading so it might not make any practical difference. :man_shrugging:

ScriabinOp8No12 replied:
Sounds good, I'll use mp3s then instead since I'm making new audio slices now anyways.

anoek commented on src/views/Lessons/Module1.tsx:
```diff
@@ -51,13 +89,30 @@ class Page1 extends Module1 {
     text(): JSX.Element | Array<JSX.Element> {
-        return (
+        return [
+            <button key="playButton" onClick={this.playAudio}>
```
If we have a play button, we probably want it to switch to "pause" or "stop" or something like that when the audio is actively playing

ScriabinOp8No12 replied:
Yep, agreed, this was something Paul and I discussed.  

I'd like to figure out how to use the useState() hook because I'm not familiar with the equivalent class component code .

Do you have any suggestions for this?

anoek replied:
Feel free to refactor to hooks if you've got a nice way of doing that, the class based stuff is still there because of inheritance stuff.

However, the class based stuff has state as well, you don't use `useState` instead you just stuff things in `this.state = { "foo": "bar" }` and update with `this.setState({"foo": "car"})`, there's still a few class based components floating around in the online-go.com code base if the reference helps: https://github.com/online-go/online-go.com/blob/devel/src/components/ModNoteModal/ModNoteModal.tsx#L39

ScriabinOp8No12 replied:
Thank you, will look into this.  I tried refactoring to use hooks, but it was quite challenging, so I read some docs on class components and implemented those instead.

anoek reviewed (commented):
Great progress!

> Also, I've updated the yarn version from 1.22.18 to 1.22.21, will that pose any problems?

Nope that's fine

ScriabinOp8No12 committed:
Trimmed audio pauses and used mp3 files instead of wav files because mp3s are smaller and better for performance, changed text animation from 500ms to 0ms since we have the audio now.

ScriabinOp8No12 committed:
Finished trimming new audio slices and using mp3s instead for module 2, fixed up module 2 page 7 to be split up and have an animation of stones being placed into the left and right side of the board

ScriabinOp8No12 committed:
Trimed audio pauses for module 3 and used mp3s instead of wav files

ScriabinOp8No12 committed:
finished module 4 audio trimming and used mp3s

ScriabinOp8No12 committed:
finished audio trimming for module 5, removed 2 white stones on module 5 page 1

ScriabinOp8No12 committed:
removed easy bot +6 from list of options, removed 6 stones as a possible handicap to start, also styled the lessons page (/learn-to-play url) with text under the buttons summarizing what the lessons are about, it's slightly off in terms of styling on both desktop and mobile, can push some pixels later lol

ScriabinOp8No12 committed:
Merge branch 'online-go:devel' into devel

ScriabinOp8No12 committed:
changed text for lesson page to be more exciting/engaging for kids, added a rounded white/blueish border for the text, properly centered it below the button

ScriabinOp8No12 committed:
Merge branch 'devel' of https://github.com/ScriabinOp8No12/kidsgoserver-update into devel

ScriabinOp8No12 committed:
Merge branch 'online-go:devel' into devel

ScriabinOp8No12 committed:
Update submodule reference for online-go.com

ScriabinOp8No12 committed:
modified and added more text on the learn-to-play page

ScriabinOp8No12 committed:
Slowed down initial stone animation on first page of module 1, also added a white stone and made it appear aroudn 10 seconds later so it aligns with the word 'followed'

ScriabinOp8No12 committed:
finished module1 timing changes

ScriabinOp8No12 committed:
slight timing modifications to module 1 and module 2, module 2 fixed up now

ScriabinOp8No12 committed:
fixed module 3 page 9, didn't realize initial state was overwriting the this.delay

ScriabinOp8No12 committed:
fixed module 3 page 6 the same way

ScriabinOp8No12 committed:
done with module 3 timing changes on stones

ScriabinOp8No12 committed:
done with stone delays for module 4 and 5, going to refactor now and make code cleaner

ScriabinOp8No12 committed:
clean up code and remove unnecessary comments

ScriabinOp8No12 commented:
Changes since adding audio (please see our discussion above):

1. Now uses mp3 files instead of wav files like you suggested
2. Trimmed and combined audio slices to have less white space
3. Changed and positioned different text on the "lessons" page to be more engaging
4. Changed various stone animations speeds
5. Updated online-go.com submodule with the 4 typescript fixes, which we discussed over email. It's not too optimal, did you get a chance to see if you could figure out the issue?  

If these changes look good, can we merge this PR in and redeploy the kidsgoserver.com and beta.kidsgoserver.com?  Paul and Terry haven't seen these recent timing changes, but they should be good to go.  We can always revert these changes if necessary.

anoek commented:
Awesome!
